browser/brave_ads/brave_ads_host.cc

```diff
@@ -0,0 +1,162 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/brave_ads/brave_ads_host.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "brave/browser/brave_ads/ads_service_factory.h"
+#include "brave/browser/brave_rewards/rewards_service_factory.h"
+#include "brave/browser/extensions/brave_component_loader.h"
+#include "brave/components/brave_ads/browser/ads_service.h"
+#include "brave/components/brave_rewards/browser/rewards_service.h"
+#include "chrome/browser/extensions/extension_service.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "content/public/browser/web_contents.h"
+#include "extensions/browser/extension_system.h"
+#include "extensions/common/constants.h"
+
+namespace brave_ads {
+
+namespace {
+
+constexpr char kAdsEnableRelativeUrl[] = "request_ads_enabled_panel.html";
+
+}  // namespace
+
+BraveAdsHost::BraveAdsHost(content::WebContents* web_contents)
+    : WebContentsObserver(web_contents) {
+  DCHECK(web_contents);
+}
+
+BraveAdsHost::~BraveAdsHost() {
+  DCHECK(!callback_);
+}
+
+void BraveAdsHost::RequestAdsEnabled(RequestAdsEnabledCallback callback) {
+  DCHECK(callback);
+
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  DCHECK(profile);
+
+  const AdsService* ads_service = AdsServiceFactory::GetForProfile(profile);
+  brave_rewards::RewardsService* rewards_service =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+  if (!rewards_service || !ads_service || !ads_service->IsSupportedLocale()) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  if (ads_service->IsEnabled()) {
+    std::move(callback).Run(true);
+    return;
+  }
+
+  if (callback_) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  Browser* browser = chrome::FindBrowserWithWebContents(web_contents());
```

Comment:
there will always be a Browser for the webcontents

Reply:
Except in tests, so perhaps we'd better keep the `if`

Reply:
Agreed, I will leave `if` here

Reply:
should have a comment then

Reply:
actually this doesn't make any sense to me. If there is no Browser for tests and we return false if there's no browser, what are we even checking here? I don't understand the purpose of this check in the first place.

Reply:
Actually there could be another case, something like AppWindow for PWA. In any case, having `if` for this method doesn't hurt and it is a common pattern. We could potentially trigger this codepath in a test that is not directly aimed at this functionality